teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+When a `Run` object is created, Tekton will validate that the `ref` is specified, and that the specified CRD type is defined, using webhook validation.
+
+After that, Tekton Pipelines expects a custom task author to implement a controller for `Run` objects that reference their type (annotated throughout this proposal with the shorthand `Run<Example>`) to take some action, and eventually update its `.status` to signal completion, either successfully or unsuccessfully, using the `conditions` model used by Tekton PipelineRuns and TaskRuns.
+
+### Validation
+
+Custom Task authors can implement webhook validation for CR objects of their provided type (e.g., to validate `Example` object definitions). Validation is optional but recommended, and examples and sample frameworks will demonstrate this functionality.
+
+### Status Reporting
+
+When the `Run<Example>` is validated and created, the Custom Task controller should be notified and begin doing some operation. When the operation begins, the controller should update the `Run`'s `.status.conditions` to report that it's ongoing:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: Unknown
+```
+
+When the operation completes, if it was successful, the condition should report `status: True`, and optionally a brief `reason` and human-readable `message`:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: True
+    reason: ExampleComplete
+    message: Yay, good times
+```
+
+If the operation was _unsuccessful_, the condition can report `status: False`, and optionally a `reason` and human-readable `message`:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: False
+    reason: ExampleFailed
+    message: Oh no bad times
+```
+
+The `Run` type's `.status` will also allow controllers to report other fields, such as `startTime`, `completionTime`, `results` (see below), and arbitrary context-dependent fields the Custom Task author wants to report. A fully-specified `Run` status might look like:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: True
+    reason: ExampleComplete
+    message: Yay, good times
+  completionTime: "2020-06-18T11:55:01Z"
+  startTime: "2020-06-18T11:55:01Z"
+  results:
+  - name: first-name
+    value: Bob
+  - name: last-name
+    value: Smith
+  arbitraryField: hello world
+  arbitraryStructuredField:
+    listOfThings: ["a", "b", "c"] 
+```
+
+### Pipeline Integration
+
+Enabling `Run`s by themselves are not terribly compelling. Their power comes from being specified in `Pipeline`s, and executed during `PipelineRun`s.
+
+Under this proposal, a user can define a Pipeline that invokes a Custom Task, specified similar to how `Task`s are specified today:
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: example-pipeline
+spec:
+  tasks:
+  ...
+  - name: example-task
+    taskRef:
+      apiVersion: example.dev/v0
+      kind: Example
+      name: my-example
```

Comment:
i want to plug a different format for these, where it's actually the entire content of the object you'd create, including objectmetadata, etc.

Check this out: https://github.com/tektoncd/pipeline/pull/2826 <-- this is adding the ability to add metadata to pipelinetasks. I think if we made pipelinetasks look more like taskruns they'd be more flexible and we'd add less add-hoc stuff like that (taskrunspecs https://github.com/tektoncd/pipeline/blob/master/docs/pipelineruns.md#specifying-taskrunspecs kind of point toward this too - but on the other hand they are at runtime so maybe providing the entire document here isn't the right balance 🤔 )

Reply:
@bobcatfish isn't it what's proposed here ?

Reply:
Yes! I think inlining sub-task specs will be _very_ cool, but I wanted to get refs in first so this doesn't expand scope too much.

Instead of:

```
apiVersion: example.dev/v0
kind: Example
metadata:
  name: my-example
spec:
  ...
---
apiVersion: tekton.dev/v1alpha1
kind: Pipeline
spec:
  tasks:
  - name: my-task
    ref:
      apiVersion: example.dev/v0
      kind: Example
      name: my-example
```

Users should eventually be able to just write:

```
apiVersion: tekton.dev/v1alpha1
kind: Pipeline
spec:
  tasks:
  - spec:  # naming is hard; task? template?
    apiVersion: example.dev/v0
    kind: Example
    spec:
      ...
```

This is nice syntactic sugar, but also avoids bugs where `Example` isn't defined first in a concatenated YAML. If I could file sub-TEPs I would :)